File: src/main/java/com/extendedae_plus/client/render/widgets/button/EAEPButton.java
package com.extendedae_plus.client.render.widgets.button;

import appeng.client.gui.Icon;
import appeng.client.gui.style.Blitter;
import appeng.client.gui.widgets.IconButton;
import net.minecraft.client.gui.GuiGraphics;
import net.minecraft.client.renderer.Rect2i;
import net.minecraft.network.chat.Component;
import net.minecraft.world.item.ItemStack;
import org.jetbrains.annotations.Nullable;

import java.util.function.Consumer;
import java.util.regex.Pattern;

public abstract class EAEPButton extends IconButton {
    protected static final Pattern PATTERN_NEW_LINE = Pattern.compile("\\n", Pattern.LITERAL);

    protected float scale = 1F;
    private final int widthOriginal;
    private final int heightOriginal;

    public EAEPButton(Consumer<EAEPButton> onPress) {
        super(button -> {
            if (!(button instanceof EAEPButton eaepButton)) return;
            onPress.accept(eaepButton);
        });
        this.widthOriginal = this.width;
        this.heightOriginal = this.height;

        this.updateTooltip();
    }

    @Override
    public void onPress() {
        super.onPress();
        this.updateTooltip();
    }

    protected void updateTooltip() {
        if (this.getNonnullAction().hasName())
            this.setMessage(this.buildMessage(
                    this.getNonnullAction().getName(),
                    this.getNonnullAction().getTooltip()));
    }

    public void setScale(float scale) {
        this.scale = scale;
        this.width = (int) Math.floor(this.widthOriginal * scale);
        this.height = (int) Math.floor(this.heightOriginal * scale);
    }

    @Override
    public Rect2i getTooltipArea() {
        var area = super.getTooltipArea();
        if (this.scale == 1F) return area;
        area.setWidth((int) Math.floor(area.getWidth() * this.scale));
        area.setHeight((int) Math.floor(area.getHeight() * this.scale));
        return area;
    }

    public abstract @Nullable EAEPActionItems getAction();

    private EAEPActionItems getNonnullAction() {
        var action = this.getAction();
        if (action == null) return EAEPActionItems.BACKING_OUT;
        else return action;
    }

    @Override
    protected Icon getIcon() {
        return this.getNonnullAction().getAEIcon();
    }

    protected Blitter getIconBlitter() {
        return this.getNonnullAction().getIconBlitter();
    }

    protected Component buildMessage(Component i18nName, @Nullable Component i18nTooltip) {
        String name = i18nName.getString();
        if (i18nTooltip == null) {
            return Component.literal(name);
        } else {
            String value = i18nTooltip.getString();
            value = PATTERN_NEW_LINE.matcher(value).replaceAll("\n");
            StringBuilder sb = new StringBuilder(value);
            int i = Math.max(sb.lastIndexOf("\n"), 0);

            while (i + 30 < sb.length() && (i = sb.lastIndexOf(" ", i + 30)) != -1) {
                sb.replace(i, i + 1, "\n");
            }

            return Component.literal(name + "\n" + sb);
        }
    }

    @Override
    public void renderWidget(GuiGraphics guiGraphics, int mouseX, int mouseY, float partial) {
        if (!this.visible) return;
        var blitter = this.getIconBlitter();
        var item = this.getItemOverlay();

        var stackPose = guiGraphics.pose();
        stackPose.pushPose();
        stackPose.translate(getX(), getY(), 0);
        stackPose.scale(this.scale, this.scale, 1.0f);
        stackPose.translate(-getX(), -getY(), 0);

        var yOffset = isHovered() ? 1 : 0;

        if (!isDisableBackground()) {
            Icon bgIcon = isHovered() ? Icon.TOOLBAR_BUTTON_BACKGROUND_HOVER
                    : isFocused() ? Icon.TOOLBAR_BUTTON_BACKGROUND_FOCUS : Icon.TOOLBAR_BUTTON_BACKGROUND;

            bgIcon.getBlitter()
                    .dest(getX() - 1, getY() + yOffset, 18, 20)
                    .zOffset(2)
                    .blit(guiGraphics);
        }
        if (item != null)
            guiGraphics.renderItem(new ItemStack(item), getX(), getY() + 1 + yOffset, 0, 3);
        else blitter.dest(getX(), getY() + 1 + yOffset).zOffset(3).blit(guiGraphics);

        stackPose.popPose();
    }

    @Override
    public void setHalfSize(boolean halfSize) {
        // HalfSize is unsupported
    }

    @Override
    public boolean isHalfSize() {
        return false;
    }
}

File: src/main/kotlin/com/extendedae_plus/client/render/widgets/button/EAEPButton.kt
package com.extendedae_plus.client.render.widgets.button

import appeng.client.gui.Icon
import appeng.client.gui.widgets.IconButton
import net.minecraft.client.gui.GuiGraphics
import net.minecraft.client.renderer.Rect2i
import net.minecraft.network.chat.Component
import net.minecraft.world.item.ItemStack
import java.util.regex.Pattern
import kotlin.math.max

abstract class EAEPButton(onPress: (EAEPButton) -> Unit) : IconButton(onPress@{
    onPress(it as? EAEPButton ?: return@onPress)
}) {
    var scale = 1f
        set(value) {
            field = value
            this.width = (this.widthOriginal * value).toInt()
            this.height = (this.heightOriginal * value).toInt()
        }
    private val widthOriginal = this.width
    private val heightOriginal = this.height

    override fun onPress() {
        super.onPress()
        this.updateTooltip()
    }

    protected fun updateTooltip() {
        val action = this.nonnullAction
        if (action.text.string.isEmpty()) return
        this.message = this.buildMessage(action.text, action.tooltip)
    }

    override fun getTooltipArea(): Rect2i {
        val area = super.getTooltipArea()
        if (this.scale == 1f) return area
        area.width = (this.widthOriginal * this.scale).toInt()
        area.height = (this.heightOriginal * this.scale).toInt()
        return area
    }

    abstract val action: EAEPActionItems?

    private val nonnullAction
        get() = this.action ?: EAEPActionItems.BackingOut

    override fun getIcon() = this.nonnullAction.aeIcon

    protected val iconBlitter
        get() = this.nonnullAction.iconBlitter

    protected fun buildMessage(i18nName: Component, i18nTooltip: Component?): Component {
        val name = i18nName.string
        if (i18nTooltip == null) {
            return Component.literal(name)
        } else {
            var value = i18nTooltip.string
            value = PATTERN_NEW_LINE.matcher(value).replaceAll("\n")
            val sb = StringBuilder(value)
            var i = max(sb.lastIndexOf("\n"), 0)

            while (i + 30 < sb.length && (sb.lastIndexOf(" ", i + 30).also { i = it }) != -1) {
                sb.replace(i, i + 1, "\n")
            }

            return Component.literal(name + "\n" + sb)
        }
    }

    override fun renderWidget(guiGraphics: GuiGraphics, mouseX: Int, mouseY: Int, partial: Float) {
        if (!this.visible) return
        val blitter = this.iconBlitter
        val item = this.itemOverlay

        val stackPose = guiGraphics.pose()
        stackPose.pushPose()
        stackPose.translate(this.x.toFloat(), this.y.toFloat(), 0f)
        stackPose.scale(this.scale, this.scale, 1.0f)
        stackPose.translate(-this.x.toFloat(), -this.y.toFloat(), 0f)

        val yOffset = if (isHovered()) 1 else 0

        if (!this.isDisableBackground) {
            val bgIcon = if (isHovered())
                Icon.TOOLBAR_BUTTON_BACKGROUND_HOVER
            else
                if (this.isFocused)
                    Icon.TOOLBAR_BUTTON_BACKGROUND_FOCUS
                else
                    Icon.TOOLBAR_BUTTON_BACKGROUND

            bgIcon.blitter
                .dest(this.x - 1, this.y + yOffset, 18, 20)
                .zOffset(2)
                .blit(guiGraphics)
        }
        if (item != null) guiGraphics.renderItem(ItemStack(item), this.x, this.y + 1 + yOffset, 0, 3)
        else blitter.dest(this.x, this.y + 1 + yOffset).zOffset(3).blit(guiGraphics)

        stackPose.popPose()
    }

    override fun setHalfSize(halfSize: Boolean) {
        // HalfSize is unsupported
    }

    override fun isHalfSize() = false

    companion object {
        protected val PATTERN_NEW_LINE: Pattern = Pattern.compile("\\n", Pattern.LITERAL)
    }
}
